controllers/remote/cluster_cache_tracker.go

```diff
@@ -374,26 +384,58 @@ func (t *ClusterCacheTracker) runningOnWorkloadCluster(ctx context.Context, c cl
 	return t.controllerPodMetadata.UID == pod.UID, nil
 }
 
-// createClient creates a cached client, and uncached client and a mapper based on a rest.Config.
-func (t *ClusterCacheTracker) createClient(ctx context.Context, config *rest.Config, cluster client.ObjectKey, indexes []Index) (client.Client, client.Client, *stoppableCache, error) {
+// createHTTPClientAndMapper creates a http client and a dynamic rest mapper for the given cluster, based on the rest.Config.
+func (t *ClusterCacheTracker) createHTTPClientAndMapper(config *rest.Config, cluster client.ObjectKey) (*http.Client, meta.RESTMapper, error) {
```

Comment:
Given that now we have this function, could we create an http client and rest mapper for each config only one and eventually use it also for the cache? Basically, `createUncachedClient` and `createCachedClient` should get those two fields as input, instead of the *rest.Config

Reply:
Well, I believe we still need the `*rest.Config` parameter.

For the `createUncachedClient` function, we need it for `client.New` call.

And, for the `createCachedClient` function, we need it for:
* `cache.New` call.
* `client.New` call.
* `newClientWithTimeout` (since it uses uses `config.Timeout` when wrapping the client before returning it).
* `go t.healthCheckCluster` goroutine.

However, it makes sense moving the `createHTTPClientAndMapper` call outside `createUncachedClient` / `createCachedClient` and adding the `httpClient` and `mapper` as parameters to the get client functions.

Reply:
Actually, I don't think that it solves anything moving the `createHTTPClientAndMapper` outside the get client functions.

We would still have to call this twice anyway:
* Before the `createUncachedClient`.
* Before `createCachedClient`. This second call is needed since the config is changed when `in-cluster` config is used, and re-using the `httpClient` and `mapper` from the `createUncachedClient` call is not what we want. 

I would let it as it is, since we cannot remove the `*rest.Config` parameter (as pointed above).

What do you think?

Reply:
I'm confused. Why can't we share the `mapper` and `httpClient` across both calls? If anything, we shouldn't duplicate these inputs for the two clients for the same cluster

Reply:
> I'm confused. Why can't we share the `mapper` and `httpClient` across both calls? If anything, we shouldn't duplicate these inputs for the two clients for the same cluster

We can share them, but it will only work only when `runningOnCluster` is `false`. The `uncachedClient` decides if the cached client with in-cluster config is needed.

If `runningOnCluster` is `true`, we need to create a different `mapper` and `httpClient` using the in-cluster config, and the previous `mapper` and `httpClient` cannot be re-used. I can update the code as suggested, but in the if branch with `runningOnCluster == true` I need override the `mapper` and `httpClient` using the in-cluster config.

Besides this, as mentioned above, we cannot remove `*rest.Config` parameter from `createCachedClient` function as suggested initially (it's used elsewhere besides calling `createHTTPClientAndMapper`). So, what I would do is adding two more extra parameters to the `createCachedClient` fuction, without removing any existing ones.

If you still believe it's better to re-use the `mapper` and `httpClient` from uncached client (but override them in the `runningOnCluster == true` if branch, I will make the change.

Reply:
See this piece of code for reference:
https://github.com/ionutbalutoiu/cluster-api/blob/8cf7a6632195f42d43da6eebb693286444e28f70/controllers/remote/cluster_cache_tracker.go#L294-L337